### PR TITLE
Space debugger vincent

### PR DESCRIPTION
**Pull request by KuuCi**
This PR implements python 3.11 support for torch nightly using flash attention 2.5.0. The docker image has passed corgi regression tests on llm-foundry as well as has been tested in an interactive using the following set up from llm foundry:
https://github.com/mosaicml/llm-foundry/pull/913

### CODE REVIEW

File: docker/Dockerfile [hunk 1]
```diff
@@ -288,16 +290,14 @@ RUN if [[ -n "$CUDA_VERSION" ]] &&  [[ -z "${PYTORCH_NIGHTLY_URL}" ]]; then \
         rm -rf /tmp/apex ; \
     fi
 
+
```
Comment (j316chuck) [inline docker/Dockerfile:293]: RM space

File: docker/Dockerfile [hunk 2]
```diff
@@ -288,16 +290,14 @@ RUN if [[ -n "$CUDA_VERSION" ]] &&  [[ -z "${PYTORCH_NIGHTLY_URL}" ]]; then \
...
 # Install Flash Attention
 ##########################
 RUN if [ -n "$CUDA_VERSION" ] ; then \
         pip${PYTHON_VERSION} install --upgrade --no-cache-dir ninja==1.11.1 && \
         pip${PYTHON_VERSION} install --upgrade --no-cache-dir --force-reinstall packaging==22.0 && \
-        git clone --branch v2.4.2 https://github.com/Dao-AILab/flash-attention.git && \
-        cd flash-attention && \
-        MAX_JOBS=1 python${PYTHON_VERSION} setup.py install && \
-        cd .. ; \
+        pip${PYTHON_VERSION} install --no-cache-dir flash-attn==2.5.0; \
```
Comment (j316chuck) [inline docker/Dockerfile:300]: I would a YAML from shashank. MPT-125m for 1b tokens.

File: docker/README.md [hunk 3]
```diff
@@ -29,18 +29,16 @@ The base flavor contains PyTorch pre-installed; the vision flavor also includes
 To install composer, once inside the image, run `pip install mosaicml`.
 
 <!-- BEGIN_PYTORCH_BUILD_MATRIX -->
-| Linux Distro   | Flavor   | PyTorch Version   | CUDA Version        | Python Version   | Docker Tags                                                                              |
```
Comment (j316chuck) [inline docker/README.md:32]: There should be no red

File: docker/generate_build_matrix.py [hunk 4]
```diff
@@ -18,7 +18,7 @@
 import tabulate
 import yaml
 
-LATEST_PYTHON_VERSION = '3.10'
+LATEST_PYTHON_VERSION = '3.11'
```
Comment (j316chuck) [inline docker/generate_build_matrix.py]: keep to python 3.10

File: docker/generate_build_matrix.py [hunk 5]
```diff
@@ -242,7 +248,25 @@ def _main():
...
+        'NVIDIA_REQUIRE_CUDA_OVERRIDE': _get_cuda_override('12.1.0'),
+        'PYTHON_VERSION': '3.11',
+        'PYTORCH_VERSION': '2.3.0',
+        'PYTORCH_NIGHTLY_URL': 'https://download.pytorch.org/whl/nightly/cu121',
+        'PYTORCH_NIGHTLY_VERSION': 'dev20240110+cu121',
+        'TAGS': ['mosaicml/pytorch:2.3.0_cu121-nightly20240110-python3.10-ubuntu20.04'],
+        'TARGET': 'pytorch_stage',
+        'TORCHVISION_VERSION': '0.18.0'
+    }
+    pytorch_entries.append(nightly_entry_311)
```
Comment (j316chuck) [inline docker/generate_build_matrix.py:268]: Double check against https://catalog.ngc.nvidia.com/orgs/nvidia/containers/pytorch/layers

File: .github/workflows/docker-configure-build-push.yaml [hunk 6]
```diff
@@ -36,6 +36,7 @@ on:
         required: true
 jobs:
   configure-build-push:
+    timeout-minutes: 1440
```
Comment (j316chuck) [inline .github/workflows/docker-configure-build-push.yaml:39]: make smaller

**Review by j316chuck — COMMENTED** [review]
LGTM barring one nit

File: .github/workflows/docker-configure-build-push.yaml (hunk 6 above)
Comment (KuuCi) [reply .github/workflows/docker-configure-build-push.yaml]: default is 6 hours so I'm just removing the line

**Review by j316chuck — APPROVED** [review]
Thanks! Please add the CORGI run and the hero run with FA2 on PyTorch 2.3.0 ✅

**Review by j316chuck — CHANGES_REQUESTED** [review]

**Review by j316chuck — APPROVED** [review]

**Review by j316chuck — APPROVED** [review]

File: docker/Dockerfile [hunk 7]
```diff
@@ -356,7 +355,8 @@ RUN apt-get update && \
 RUN pip install --no-cache-dir --upgrade \
         certifi${CERTIFI_VERSION} \
         ipython${IPYTHON_VERSION} \
-        urllib3${URLLIB3_VERSION}
+        urllib3${URLLIB3_VERSION} \
+        python-snappy
```
Comment (mvpatel2000) [inline docker/Dockerfile:359]: why?

File: docker/Dockerfile [hunk 8]
```diff
@@ -269,6 +270,7 @@ RUN if [ -n "$MOFED_VERSION" ] ; then \
         rm -rf /tmp/mofed ; \
     fi
 
+
```
Comment (mvpatel2000) [inline docker/Dockerfile:273]:
```suggestion
```

File: docker/Dockerfile [hunk 9]
```diff
@@ -118,6 +118,7 @@ RUN apt-get update && \
         tcl \
         libjpeg8-dev \
         less \
+        libsnappy-dev \
```
Comment (mvpatel2000) [inline docker/Dockerfile:121]: why?

File: docker/generate_build_matrix.py [hunk 10]
```diff
@@ -242,7 +242,25 @@ def _main():
...
+        'NVIDIA_REQUIRE_CUDA_OVERRIDE': _get_cuda_override('12.1.0'),
+        'PYTHON_VERSION': '3.11',
+        'PYTORCH_VERSION': '2.3.0',
+        'PYTORCH_NIGHTLY_URL': 'https://download.pytorch.org/whl/nightly/cu121',
+        'PYTORCH_NIGHTLY_VERSION': 'dev20240110+cu121',
+        'TAGS': ['mosaicml/pytorch:2.3.0_cu121-nightly20240110-python3.11-ubuntu20.04'],
+        'TARGET': 'pytorch_stage',
+        'TORCHVISION_VERSION': '0.18.0'
+    }
+    pytorch_entries.append(nightly_entry_311)
```
Comment (mvpatel2000) [inline docker/generate_build_matrix.py:262]: Let's not add -- we only need one version of nightly for testing and it won't be used in CI/CD

**Review by mvpatel2000 — COMMENTED** [review]
Is this planned on being merged? or no?

File: docker/Dockerfile (hunk 7 above)
Comment (KuuCi) [reply docker/Dockerfile:359]: python snappy doesn't install correctly without first apt install and then force pip install

File: docker/generate_build_matrix.py (hunk 10 above)
Comment (KuuCi) [reply docker/generate_build_matrix.py:262]: So which versions should we be supporting with 3.11 ~~ 2.1.2?

File: docker/Dockerfile [hunk 11]
```diff
@@ -294,10 +296,7 @@ RUN if [[ -n "$CUDA_VERSION" ]] &&  [[ -z "${PYTORCH_NIGHTLY_URL}" ]]; then \
 RUN if [ -n "$CUDA_VERSION" ] ; then \
         pip${PYTHON_VERSION} install --upgrade --no-cache-dir ninja==1.11.1 && \
         pip${PYTHON_VERSION} install --upgrade --no-cache-dir --force-reinstall packaging==22.0 && \
-        git clone --branch v2.4.2 https://github.com/Dao-AILab/flash-attention.git && \
-        cd flash-attention && \
-        MAX_JOBS=1 python${PYTHON_VERSION} setup.py install && \
-        cd .. ; \
+        MAX_JOBS=1 pip${PYTHON_VERSION} install --no-cache-dir flash-attn==2.5.0; \
```
Comment (mvpatel2000) [inline docker/Dockerfile:299]: What does MAX_JOBS=1 do?

File: docker/generate_build_matrix.py (hunk 10 above)
Comment (mvpatel2000) [reply docker/generate_build_matrix.py:262]: JK

**Review by mvpatel2000 — APPROVED** [review]

File: docker/Dockerfile (hunk 11 above)
Comment (KuuCi) [reply docker/Dockerfile:299]: If you don't specify MAX_JOBS=1 it'll default to some higher number and cause an out of mem error

File: docker/Dockerfile (hunk 9 above)
Comment (KuuCi) [reply docker/Dockerfile:121]: same as above

**Pull request merged**